### PR TITLE
feat: add flag to skip deleting obsolete referrers index

### LINKED ISSUE #915: Improve the error msg for OCI v1.0 registries or registries that don't support deletion API
### What happened in your environment?

When using ORAS CLI to interact with OCI v1.0 compliant registries such as GHCR or registries that [don't support deletion API](https://github.com/docker/hub-feedback/issues/1759) (e.g. Docker Hub), ORAS will return an unexpected error message and may mislead or confuse users. It happens when attaching an artifact to the image for the second time.

### What did you expect to happen?

Improve the error message to be more descriptive and user-friendly, and tell users about the right result exactly. Giving an example: This artifact has been attached to xxx. ORAS can't delete the referrers index since this is an OCI v1.0 compliant registry.

### How can we reproduce it?

I have attached a file to the sample Alpine image in GHCR. Then I attached another file to the same image and it returns the error as follows:

```
oras attach --artifact-type=sbom/test ghcr.io/feynmanzhou/alpine:latest sbom.json -u feynmanzhou --password-stdin
Uploading 57761fa432ee sbom.json
Uploaded  57761fa432ee sbom.json
Error: failed to delete dangling referrers index sha256:de0b503017452ba6b070981414bfc31f8e6096852a36ff4f3223fdc0ff411566 for referrers tag sha256-1304f174557314a7ed9eddb4eab12fed12cb0cd9809e4c28f29af86979a3c870: DELETE "https://ghcr.io/v2/feynmanzhou/alpine/manifests/sha256:de0b503017452ba6b070981414bfc31f8e6096852a36ff4f3223fdc0ff411566": response status code 405: unsupported: The operation is unsupported.
```

It may confuse users why this `attach` behavior failed due to an old index manifest deletion failure. In fact, this file has been attached to the image successfully but it doesn't tell users the right result unless users verify it using `oras discover` as below.  

```
oras discover ghcr.io/feynmanzhou/alpine:latest -o tree -u feynmanzhou --password-stdin
ghcr.io/feynmanzhou/alpine@sha256:1304f174557314a7ed9eddb4eab12fed12cb0cd9809e4c28f29af86979a3c870
├── application/vnd.cncf.notary.signature
│   └── sha256:85f65b60a6e260d52ba5f8d556d444320b92c2ac140ec8d4944dbd8b03deb18c
└── sbom/test
    └── sha256:ee375fb1111ca3d0c60663115e2cdce3e7061821ab26059852ecc43ceb8ee8cd


### What is the version of your ORAS CLI?

```
Version:        1.0.0
Go version:     go1.20.2
Git commit:     b58e7b910ca556973d111e9bd734a71baef03db2
Git tree state: clean
```

### What is your OS environment?

Ubuntu 20.08

### Are you willing to submit PRs to fix it?

- [ ] Yes, I am willing to fix it.

### LINKED ISSUE #952: make `oras cp` skip dangling referrers index deletion error
### What is the version of your ORAS CLI

1.0.0

### What would you like to be added?

`oras cp` is by-design to stop on the first-met error. if the copy destination is an OCI-1.0-compliant registry with deletion disabled, then `oras cp` operation might be done half-way without copying all the nodes.

Consider copying the below artifacts via index `A` to such a registry(OCI 1.0+delete disabled) with concurrency set to 1.
```mermaid
graph TD;
  A--platform b-->B
  A--platform c-->C
  A--platform d-->D
  B--subject-->E
  C--subject-->E
  D--subject-->E
```
The copy process will cover all referrers of `E`, since the concurrency is set to 1, those three referrers, `B`, `C` and `D` will be copied one by one. Let's assume the sequence is `B` -> `C` -> `D`:
1) When copying `B`, the referrer index is created with one descriptor pointing to `B`
2) When copying `C`, the referrer index created in step 1) is fetched and append another descriptor of `B`. The referrer index created in step 1) is dangling and need to be cleaned up
3) clean up will fail and copy terminated

In this case, `D` is not covered and the copy process stops.

### Why is this needed for ORAS?

Copying progress should not be terminated when such failure encountered, it is not terminal and user can clean up the dangling referrer index after the cp process.

### Are you willing to submit PRs to contribute to this feature?

- [X] Yes, I am willing to implement it.

### LINKED ISSUE #954: allow oras command to skip referrer index clean up
### What is the version of your ORAS CLI

1.0.0

### What would you like to be added?

To all oras commands causing referrers update, a new flag `--gc` should be added to skip deleting the dangling referrers index:
- By default, `--gc` is not set, dangling referrers index will not be clean
- If `--gc` is set, clean dangling referrers index when needed

This flag should be added to below commands
- oras attach
- oras cp
- oras manifest push
- oras manifest delete

### Why is this needed for ORAS?

Referrer index is the alternative that OCI-1.0 compliant registries can utilize to support artifact reference types, see [OCI spec 1.0-rc.2](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc2/spec.md#unavailable-referrers-api).

If a referrers index `i` exists for a certain subject artifact `a`. When adding or removing referrers for `a`, content of `i` will be updated and a new referrers index will be created and retagged following the [tag schema](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc2/spec.md#referrers-tag-schema). The index `i` contains obsolete content without any tag referenced. 

When running commands like `oras attach` and `oras cp`, it's possible that the used token scope doesn't cover delete, or the registry might disable deletion, so it's better to provide oras an option to skip cleaning the dangling referrers index.
 
Also checked OCI spec, cleanup is optional for registry clients.

### Are you willing to submit PRs to contribute to this feature?

- [X] Yes, I am willing to implement it.

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PR adds a flag to `oras attach`, `oras cp` and `oras manifest push`. The flag can be used to specify whether to clean up old referrers index when referrers API is not available.

There is not a breaking change and by default, mentioned commands will still do referrers cleanup.

Depends on merging of #967.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #954, fixes #952, fixes #915

**Please check the following list**:
- [ ]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: cmd/oras/internal/option/referrers.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+// Referrers option struct.
+type Referrers struct {
+	SkipDeleteReferrers bool
+}
+
+// ApplyFlags applies flags to a command flag set.
+func (opts *Referrers) ApplyFlags(fs *pflag.FlagSet) {
+	fs.BoolVarP(&opts.SkipDeleteReferrers, "skip-delete-referrers", "", false, "skip deleting old referrers index, only work on registry when referrers API is not supported")
```
Comment (Wwwsylvia) [inline cmd/oras/internal/option/referrers.go:32]: Question: So ORAS will still delete the dangling referrers index by default as before? Is it a breaking change? My understanding was that we would skip the deletion by default and would have something like `--gc` to enable deletion?

Comment (qweeah) [reply cmd/oras/internal/option/referrers.go:31]: >  So ORAS will still delete the dangling referrers index by default as before?

Yes. By default oras **will** delete the dangling referrers index as before.

File: cmd/oras/internal/option/referrers.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+// SetReferrersGC sets the referrers GC option for the passed-in target.
+func (opts *Referrers) SetReferrersGC(target any, logger logrus.FieldLogger) {
+	if repo, ok := target.(*remote.Repository); ok {
+		repo.SkipReferrersGC = opts.SkipDeleteReferrers
+	} else if !opts.SkipDeleteReferrers {
+		logger.Warnln("referrers deletion can only be enforced upon registry")
+	}
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/referrers.go:40]: In the scenario of oci layout folder, it seems that this warning is always printed unexpectedly.

Comment (shizhMSFT) [reply cmd/oras/internal/option/referrers.go:40]: Is that `opts.SkipDeleteReferrers` instead of `!opts.SkipDeleteReferrers`?

**Review by shizhMSFT — COMMENTED** [review]
What about the manifest delete path?

File: cmd/oras/internal/option/referrers.go (hunk 2 above)
Comment (qweeah) [reply cmd/oras/internal/option/referrers.go:40]: Yes you are right, otherwise by default it will always be printed out.

**Comment by qweeah** [comment]
> What about the manifest delete path?

Referrers index deletion happens after manifest deletion right? If the registry already supports deletion and the token used covers manifest deletion, why would referrers index fail?

**Comment by shizhMSFT** [comment]
> Referrers index deletion happens after manifest deletion right? If the registry already supports deletion and the token used covers manifest deletion, why would referrers index fail?

LoL. Makes sense.

File: cmd/oras/internal/option/referrers.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	fs.BoolVarP(&opts.SkipDeleteReferrers, "skip-delete-referrers", "", false, "skip deleting old referrers index, only work on registry when referrers API is not supported")
+}
+
+// SetReferrersGC sets the referrers GC option for the passed-in target.
+func (opts *Referrers) SetReferrersGC(target any, logger logrus.FieldLogger) {
+	if repo, ok := target.(*remote.Repository); ok {
+		repo.SkipReferrersGC = opts.SkipDeleteReferrers
+	} else if opts.SkipDeleteReferrers {
+		// not a registry, can't skip referrers deletion
+		// TODO: change default value in 2.0
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/referrers.go:40]: Let's remove this comment. Instead, we should create an issue to discuss this. We probably don't want to change the default value but change the flag name.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**